File: alpha/experiments/item-storage-protocol-standalone/src/crypto.ts
// The cryptographic floor: fingerprints (SHA-256) and signatures (Ed25519),
// both from Node's built-in crypto so the whole suite has zero dependencies.
//
// Keys are derived deterministically from a master seed + a label, so every run
// produces the same keypairs, the same signatures (Ed25519 is deterministic per
// RFC 8032), and therefore byte-identical ledgers. That reproducibility is what
// lets every assertion be exact.
//
// SEAM: hex-encoded SHA-256 stands in for a CIDv1 over DAG-CBOR. Real atproto
// content addresses are multihash-tagged CIDs; here a bare hex digest is the
// "fingerprint". The tamper-evidence property is identical; only the encoding
// differs.

import {
  createHash,
  createPrivateKey,
  createPublicKey,
  sign as nodeSign,
  verify as nodeVerify,
  type KeyObject,
} from "node:crypto";
import { canonicalize } from "./canonical.ts";

// PKCS8 DER prefix for an Ed25519 private key, followed by the 32-byte seed.
// This is a fixed ASN.1 structure (SEQUENCE / version / Ed25519 OID / OCTET
// STRING), so prefix ++ seed is a valid PKCS8 key we can import deterministically.
const PKCS8_ED25519_PREFIX = Buffer.from(
  "302e020100300506032b657004220420",
  "hex",
);

/** SHA-256 of a UTF-8 string or buffer, hex-encoded. This is our "fingerprint". */
export function sha256hex(data: string | Buffer): string {
  return createHash("sha256").update(data).digest("hex");
}

/** Fingerprint of raw item bytes. Named to match the Part 1 vocabulary. */
export function fingerprint(bytes: Buffer): string {
  return sha256hex(bytes);
}

/** Hash of a value's canonical form — the thing signatures are actually taken over. */
export function hashCanonical(value: unknown): string {
  return sha256hex(canonicalize(value));
}

export type Keypair = {
  label: string;
  privateKey: KeyObject;
  publicKey: KeyObject;
  /** Raw 32-byte Ed25519 public key, hex-encoded. The pinnable identity. */
  publicKeyHex: string;
};

/** Derive a deterministic Ed25519 keypair from a master seed and a role label. */
export function deriveKeypair(masterSeed: string, label: string): Keypair {
  const seed = createHash("sha256")
    .update(`${masterSeed}::keyseed::${label}`)
    .digest(); // 32 bytes, the Ed25519 secret scalar seed
  const der = Buffer.concat([PKCS8_ED25519_PREFIX, seed]);
  const privateKey = createPrivateKey({ key: der, format: "der", type: "pkcs8" });
  const publicKey = createPublicKey(privateKey);
  const jwk = publicKey.export({ format: "jwk" }) as { x: string };
  const publicKeyHex = Buffer.from(jwk.x, "base64url").toString("hex");
  return { label, privateKey, publicKey, publicKeyHex };
}

/** Reconstruct a verify-only KeyObject from a pinned raw public key hex. */
export function publicKeyFromHex(publicKeyHex: string): KeyObject {
  const jwk = {
    kty: "OKP",
    crv: "Ed25519",
    x: Buffer.from(publicKeyHex, "hex").toString("base64url"),
  };
  return createPublicKey({ key: jwk, format: "jwk" });
}

/** Sign a UTF-8 message, returning a hex signature. */
export function signMessage(kp: Keypair, message: string): string {
  return nodeSign(null, Buffer.from(message, "utf8"), kp.privateKey).toString("hex");
}

/** Verify a hex signature against a pinned public key (hex or KeyObject). */
export function verifyMessage(
  publicKey: string | KeyObject,
  message: string,
  signatureHex: string,
): boolean {
  const key = typeof publicKey === "string" ? publicKeyFromHex(publicKey) : publicKey;
  try {
    return nodeVerify(
      null,
      Buffer.from(message, "utf8"),
      key,
      Buffer.from(signatureHex, "hex"),
    );
  } catch {
    return false;
  }
}

File: alpha/experiments/item-store/src/crypto.rs
//! The cryptographic floor: fingerprints (SHA-256) and signatures (Ed25519).
//!
//! Ports `item-storage-protocol-standalone/src/crypto.ts`. Keys are derived
//! deterministically from a master seed + a role label, so every run produces
//! the same keypairs and (Ed25519 being deterministic per RFC 8032) the same
//! signatures — the reproducibility that lets every assertion be exact.
//!
//! `SEAM:` a hex-encoded SHA-256 stands in for a `CIDv1` over `DAG-CBOR` (Phase 2
//! closes this with the in-corpus `serde_ipld_dagcbor` + `ipld-core` + `sha2`
//! path). The tamper-evidence property is identical; only the encoding differs.

use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use sha2::{Digest, Sha256};
use zeroize::Zeroize;

/// The length in bytes of a raw Ed25519 public key.
const PUBLIC_KEY_LEN: usize = 32;

/// Errors from reconstructing a public key from its wire (hex) form.
#[derive(Debug, thiserror::Error)]
pub enum CryptoError {
    /// The supplied string was not valid hexadecimal.
    #[error("public key is not valid hexadecimal")]
    InvalidHex,
    /// The decoded key was not exactly [`PUBLIC_KEY_LEN`] bytes.
    #[error("public key must be {PUBLIC_KEY_LEN} bytes, got {got}")]
    InvalidKeyLength {
        /// The number of bytes actually decoded.
        got: usize,
    },
    /// The bytes did not decode to a valid Ed25519 curve point.
    #[error("public key is not a valid Ed25519 point")]
    InvalidPublicKey,
}

/// SHA-256 of raw bytes, hex-encoded. This is our "fingerprint".
#[must_use]
pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hex::encode(hasher.finalize())
}

/// A keypair bound to a role label.
///
/// Holds the Ed25519 signing key, which is zeroized on drop (via the
/// `ed25519-dalek` `zeroize` feature). `Debug` is intentionally not derived so
/// the secret cannot leak through formatting.
pub struct Keypair {
    label: String,
    signing_key: SigningKey,
}

impl Keypair {
    /// The role label this keypair was derived for (e.g. `"customer"`).
    #[must_use]
    pub fn label(&self) -> &str {
        &self.label
    }

    /// The public (verifying) key — the pinnable identity.
    #[must_use]
    pub fn verifying_key(&self) -> VerifyingKey {
        self.signing_key.verifying_key()
    }

    /// The raw 32-byte public key, hex-encoded — the pinnable identity on the wire.
    #[must_use]
    pub fn public_key_hex(&self) -> String {
        hex::encode(self.verifying_key().to_bytes())
    }

    /// Sign a UTF-8 message, returning a hex-encoded signature.
    #[must_use]
    pub fn sign_message(&self, message: &str) -> String {
        let signature = self.signing_key.sign(message.as_bytes());
        hex::encode(signature.to_bytes())
    }
}

/// Derive a deterministic Ed25519 keypair from a master seed and a role label.
///
/// The seed for the key is `SHA-256("{master_seed}::keyseed::{label}")`, matching
/// the TypeScript oracle so signatures are comparable against it.
#[must_use]
pub fn derive_keypair(master_seed: &str, label: &str) -> Keypair {
    let mut hasher = Sha256::new();
    hasher.update(format!("{master_seed}::keyseed::{label}").as_bytes());
    let mut seed: [u8; 32] = hasher.finalize().into();
    let signing_key = SigningKey::from_bytes(&seed);
    seed.zeroize();
    Keypair {
        label: label.to_owned(),
        signing_key,
    }
}

/// Reconstruct a verify-only public key from its raw hex form (as a peer pins it).
///
/// # Errors
///
/// Returns [`CryptoError`] if the input is not valid hex, is not
/// [`PUBLIC_KEY_LEN`] bytes, or is not a valid Ed25519 point.
pub fn public_key_from_hex(public_key_hex: &str) -> Result<VerifyingKey, CryptoError> {
    let bytes = hex::decode(public_key_hex).map_err(|_| CryptoError::InvalidHex)?;
    let array: [u8; PUBLIC_KEY_LEN] = bytes
        .as_slice()
        .try_into()
        .map_err(|_| CryptoError::InvalidKeyLength { got: bytes.len() })?;
    VerifyingKey::from_bytes(&array).map_err(|_| CryptoError::InvalidPublicKey)
}

/// Verify a hex signature over a UTF-8 message against a pinned public key.
///
/// Returns `false` for a bad signature, a malformed signature encoding, or a
/// mismatched key — verification failure is a legitimate `false`, not an error
/// to propagate (matching the oracle's semantics).
#[must_use]
pub fn verify_message(verifying_key: &VerifyingKey, message: &str, signature_hex: &str) -> bool {
    let Ok(signature_bytes) = hex::decode(signature_hex) else {
        return false;
    };
    let Ok(signature) = Signature::from_slice(&signature_bytes) else {
        return false;
    };
    verifying_key.verify(message.as_bytes(), &signature).is_ok()
}

#[cfg(test)]
mod tests {
    use super::{derive_keypair, public_key_from_hex, verify_message, CryptoError};

    #[test]
    fn derivation_is_deterministic_in_seed_and_label() {
        let a = derive_keypair("master", "customer");
        let b = derive_keypair("master", "customer");
        assert_eq!(a.public_key_hex(), b.public_key_hex());
    }

    #[test]
    fn distinct_labels_derive_distinct_keys() {
        let customer = derive_keypair("master", "customer");
        let provider = derive_keypair("master", "provider");
        assert_ne!(customer.public_key_hex(), provider.public_key_hex());
    }

    #[test]
    fn public_key_hex_round_trips() {
        let kp = derive_keypair("master", "customer");
        let reconstructed = public_key_from_hex(&kp.public_key_hex()).expect("valid hex");
        assert_eq!(reconstructed.to_bytes(), kp.verifying_key().to_bytes());
    }

    #[test]
    fn malformed_public_key_hex_is_rejected() {
        assert!(matches!(
            public_key_from_hex("not-hex"),
            Err(CryptoError::InvalidHex)
        ));
        assert!(matches!(
            public_key_from_hex("00ff"),
            Err(CryptoError::InvalidKeyLength { got: 2 })
        ));
    }

    #[test]
    fn signature_over_one_message_does_not_verify_another() {
        let kp = derive_keypair("master", "customer");
        let key = public_key_from_hex(&kp.public_key_hex()).expect("valid hex");
        let signature = kp.sign_message("hello");
        assert!(verify_message(&key, "hello", &signature));
        assert!(!verify_message(&key, "hell0", &signature));
    }

    #[test]
    fn garbage_signature_encoding_returns_false_not_panic() {
        let kp = derive_keypair("master", "customer");
        let key = public_key_from_hex(&kp.public_key_hex()).expect("valid hex");
        assert!(!verify_message(&key, "hello", "zz"));
    }
}
